src/data/bucket/symbol_bucket.js

```diff
@@ -752,6 +761,78 @@ class SymbolBucket implements Bucket {
         }
         return collisionArrays;
     }
+
+    sortFeatures(angle: number) {
+        if (!this.sortFeaturesByY) return;
+
+        if (this.sortedAngle === angle) return;
+        this.sortedAngle = angle;
+
+        // If the symbols are allowed to overlap sort them by their vertical screen position.
+        // The index array buffer is rewritten to reference the (unchanged) vertices in the
+        // sorted order.
+
+        // To avoid sorting the actual symbolInstance array we sort an array of indexes.
+        const symbolInstanceIndexes = [];
+        for (let i = 0; i < this.symbolInstances.length; i++) {
+            symbolInstanceIndexes.push(i);
+        }
+
+        const sin = Math.sin(angle),
+            cos = Math.cos(angle);
+
+        symbolInstanceIndexes.sort((aIndex, bIndex) => {
+            const a = this.symbolInstances[aIndex];
+            const b = this.symbolInstances[bIndex];
+            const aRotated = (sin * a.anchor.x + cos * a.anchor.y) | 0;
+            const bRotated = (sin * b.anchor.x + cos * b.anchor.y) | 0;
+            return (aRotated - bRotated) || (b.featureIndex - a.featureIndex);
+        });
+
+        this.text.indexArray.clear();
+        this.icon.indexArray.clear();
+
+        // In most cases this outer loop over the segments doesn't do anything.
+        // Usually there is just one segment. But sometimes vertices spill over into
+        // additional segments. In this case we need to make sure that vertices
+        // are referenced from their matching index array segment. This is done by
+        // adding all the indexes from the first segment, then the second, etc.
+        const numTextSegments = this.text.segments.get().length;
+        for (let segmentIndex = 0; segmentIndex < numTextSegments; segmentIndex++) {
+            for (const i of symbolInstanceIndexes) {
+                const symbolInstance = this.symbolInstances[i];
+
+                for (const placedTextSymbolIndex of symbolInstance.placedTextSymbolIndices) {
+                    const placedSymbol = this.placedGlyphArray.get(placedTextSymbolIndex);
```

Comment:
Am I right in reading this as "we sort symbols within a segment, but symbols can't reorder between segments"? And if that's the case, aren't we basically giving up on sorting when there are multiple segments? That seems OK since this is an edge case, but if we're going to give up, why not just early-exit the entire sorting process?

Reply:
Yep, it's kind of giving up. I hadn't thought about an early-exit but that's probably cleaner. This approach sorts them so that they look mostly right, but is more complicated and probably error-prone. I'll switch to an early exit!